src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php

```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
```

Comment:
@stof: Looking at the original code again, I think the restart step doesn't do anything anyway. Aliases are replaced only if the definition of their target is marked as private; the process is to mark the definition as public and replace all references to the alias target with references to the alias id (and then removing the definition of the alias target). This process does not introduce any new private definitions, therefore no alias target that was previously seen as having a public definition should have a private one after it. Am I missing something?